sql-tuning-best-practice.md

```diff
@@ -0,0 +1,954 @@
+
+In TiDB, leveraging indexes effectively is crucial for performance tuning, as it can significantly reduce the number of KV [RPC](/glossary.md#remote-procedure-call-rpc) requests. By minimizing these requests, you can greatly improve query performance and overall system efficiency. Here are some key strategies:
+
+- Avoiding full table scans
+- Avoiding sorting
+- Skipping row lookups when possible
+
+This section showcases three practical examples that illustrate effective indexing strategies in TiDB, focusing on the use of composite and covered indexes for both filtering and sorting operations.
+
+### SQL Tuning with a Covered Index
+
+A covered index is designed to include all columns referenced in the filter and select clauses. The query below requires an index lookup of 2597411 rows, taking 46.4 seconds to execute. TiDB needs to dispatch 67 cop tasks for the index range scan on logs_idx, identified as `IndexRangeScan_11`, and 301 cop tasks for table access via `TableRowIDScan_12`. By utilizing a covered index, the index lookup can be avoided, leading to improved performance.
+
+```sql
+SELECT
+  SUM(`logs`.`amount`)
+FROM
+  `logs`
+WHERE
+  `logs`.`user_id` = 1111
+  AND `logs`.`snapshot_id` IS NULL
+  AND `logs`.`status` IN ('complete', 'failure')
+  AND `logs`.`source_type` != 'online'
+  AND (
+    `logs`.`source_type` IN ('user', 'payment')
+    OR `logs`.`source_type` IN (
+      'bank_account',
+    )
+    AND `logs`.`target_type` IN ('bank_account')
+  );
+```
+
+```
++-------------------------------+------------+---------+-----------+--------------------------------------------------------------------------+------------------------------------------------------------+
+| id                            | estRows    | actRows | task      | access object                                                            | execution info                                             | 
++-------------------------------+------------+---------+-----------+--------------------------------------------------------------------------+------------------------------------------------------------+
+| HashAgg_18                   | 1.00       | 2571625.22 | 1       | root      |                                                              | time:46.4s, loops:2, partial_worker:{wall_time:46.37,   ...|
+| └─IndexLookUp_19             | 1.00       | 2570096.68 | 301     | root      |                                                              | time:46.4s, loops:2, index_task: {total_time: 45.8s,    ...|
+|   ├─IndexRangeScan_11(Build) | 1309.50    | 317033.98  | 2597411 | cop[tikv] | table:logs, index:logs_idx(snapshot_id, user_id, status)     | time:228ms, loops:2547, cop_task: {num: 67, max: 2.17s, ...|
```

Comment:
Should we run `ANALYZE TABLE logs` first, or is this difference between `estRows` and `actRows` correct?